engine/src/main/java/org/n52/javaps/engine/impl/EngineImpl.java

```diff
@@ -182,20 +192,25 @@ public JobId execute(OwsCode identifier,
 
     private Result onJobCompletion(Job job) throws EngineException {
         try {
+            this.jobStatusLock.writeLock().lock();
             this.cancelers.remove(job.getJobId());
             this.resultPersistence.save(job);
             this.jobs.remove(job.getJobId());
             return this.resultPersistence.getResult(job.getJobId());
         } finally {
             job.destroy();
+            this.jobStatusLock.writeLock().unlock();
         }
     }
 
     @Override
     public Future<Result> getResult(JobId identifier) throws JobNotFoundException {
         LOG.info("Getting result {}", identifier);
+
+        this.jobStatusLock.readLock().lock();
         Job job = this.jobs.get(identifier);
```

Comment:
Same as above, if `this.jobs.get(identifier);`, fails, the lock is never unlocked.